apps/envited-x-data-space/modules/AddAssets/AddAssets.tsx

```diff
@@ -46,16 +48,73 @@ export const AddAssets = () => {
 
       const filesArray = Array.from(data.assets as FileList)
 
+      filesArray.forEach((file, index) => {
+        setUploadAssetsState(prev => {
+          const updated = [...prev]
+          updated[index] = {
+            ...updated[index],
+            progress: 0,
+            status: UploadAssetStatus.queued,
+          }
+
+          return updated
+        })
+      })
+
       const processFiles = pipe(map(processFile), Promise.all.bind(Promise))
       const filesData = await processFiles(filesArray)
       const uploadData = await validateAndUploadAssets(filesData as AssetFile[])
-      const uploadResults = await Promise.all(map(file => uploadFile(filesArray, file), uploadData))
 
-      map(({ success, file }: { success: boolean; file: string }) =>
-        success ? successNotification(`${file} successfully uploaded`) : error(`${file} already exists`),
-      )(uploadResults)
+      const uploadPromises = uploadData.map(async (file, index) => {
```

Comment:
Can this be abstracted out into a reusable and testable upload flow that emits status updates?